panic: runtime error: invalid memory address or nil pointer dereference

/kind bug

**What steps did you take and what happened:**
`kubectl delete cluster capi-quickstart`

output of control manager is:
```"="capi-quickstart-md-0-5fd4j" "namespace"="default" "openStackCluster"="capi-quickstart" "openStackMachine"="capi-quickstart-md-0-5fd4j" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="OpenStackMachine"

I0615 07:39:49.553344       1 http.go:97] controller-runtime/webhook/webhooks "msg"="received request" "webhook"="/mutate-infrastructure-cluster-x-k8s-io-v1alpha5-openstackmachine" "UID"="2d0d8970-f57a-4c17-8295-7f9fb92a3a13" "kind"={"group":"infrastructure.cluster.x-k8s.io","version":"v1alpha5","kind":"OpenStackMachine"} "resource"={"group":"infrastructure.cluster.x-k8s.io","version":"v1alpha5","resource":"openstackmachines"}

I0615 07:39:49.557673       1 http.go:136] controller-runtime/webhook/webhooks "msg"="wrote response" "code"=200 "reason"="" "webhook"="/mutate-infrastructure-cluster-x-k8s-io-v1alpha5-openstackmachine" "UID"="2d0d8970-f57a-4c17-8295-7f9fb92a3a13" "allowed"=true

I0615 07:39:49.558956       1 http.go:97] controller-runtime/webhook/webhooks "msg"="received request" "webhook"="/mutate-infrastructure-cluster-x-k8s-io-v1alpha5-openstackmachine" "UID"="f26e2a56-3163-4a64-860c-f319013ea1e8" "kind"={"group":"infrastructure.cluster.x-k8s.io","version":"v1alpha5","kind":"OpenStackMachine"} "resource"={"group":"infrastructure.cluster.x-k8s.io","version":"v1alpha5","resource":"openstackmachines"}

I0615 07:39:49.560535       1 http.go:136] controller-runtime/webhook/webhooks "msg"="wrote response" "code"=200 "reason"="" "webhook"="/mutate-infrastructure-cluster-x-k8s-io-v1alpha5-openstackmachine" "UID"="f26e2a56-3163-4a64-860c-f319013ea1e8" "allowed"=true

I0615 07:39:49.561537       1 http.go:97] controller-runtime/webhook/webhooks "msg"="received request" "webhook"="/mutate-infrastructure-cluster-x-k8s-io-v1alpha5-openstackmachine" "UID"="73845b83-1978-4c8b-9e34-000de8227ce8" "kind"={"group":"infrastructure.cluster.x-k8s.io","version":"v1alpha5","kind":"OpenStackMachine"} "resource"={"group":"infrastructure.cluster.x-k8s.io","version":"v1alpha5","resource":"openstackmachines"}

I0615 07:39:49.561947       1 http.go:136] controller-runtime/webhook/webhooks "msg"="wrote response" "code"=200 "reason"="" "webhook"="/mutate-infrastructure-cluster-x-k8s-io-v1alpha5-openstackmachine" "UID"="73845b83-1978-4c8b-9e34-000de8227ce8" "allowed"=true

I0615 07:39:49.581094       1 openstackmachine_controller.go:287] controller/openstackmachine "msg"="Cluster infrastructure is not ready yet, requeuing machine" "cluster"="capi-quickstart" "machine"="capi-quickstart-md-0-7674b8c4d7-m7sbl" "name"="capi-quickstart-md-0-nxmsz" "namespace"="default" "openStackCluster"="capi-quickstart" "openStackMachine"="capi-quickstart-md-0-nxmsz" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="OpenStackMachine"

I0615 07:39:49.583766       1 http.go:97] controller-runtime/webhook/webhooks "msg"="received request" "webhook"="/validate-infrastructure-cluster-x-k8s-io-v1alpha5-openstackmachine" "UID"="aa2e6699-ad6c-4cd4-b587-0c15323b7aec" "kind"={"group":"infrastructure.cluster.x-k8s.io","version":"v1alpha5","kind":"OpenStackMachine"} "resource"={"group":"infrastructure.cluster.x-k8s.io","version":"v1alpha5","resource":"openstackmachines"}

I0615 07:39:49.584716       1 http.go:136] controller-runtime/webhook/webhooks "msg"="wrote response" "code"=200 "reason"="" "webhook"="/validate-infrastructure-cluster-x-k8s-io-v1alpha5-openstackmachine" "UID"="aa2e6699-ad6c-4cd4-b587-0c15323b7aec" "allowed"=true

I0615 07:39:49.587157       1 http.go:97] controller-runtime/webhook/webhooks "msg"="received request" "webhook"="/validate-infrastructure-cluster-x-k8s-io-v1alpha5-openstackmachine" "UID"="97feeb40-5b30-4e6a-b8b0-a5df301db5f8" "kind"={"group":"infrastructure.cluster.x-k8s.io","version":"v1alpha5","kind":"OpenStackMachine"} "resource"={"group":"infrastructure.cluster.x-k8s.io","version":"v1alpha5","resource":"openstackmachines"}

I0615 07:39:49.587908       1 http.go:136] controller-runtime/webhook/webhooks "msg"="wrote response" "code"=200 "reason"="" "webhook"="/validate-infrastructure-cluster-x-k8s-io-v1alpha5-openstackmachine" "UID"="97feeb40-5b30-4e6a-b8b0-a5df301db5f8" "allowed"=true

I0615 07:39:49.589089       1 http.go:97] controller-runtime/webhook/webhooks "msg"="received request" "webhook"="/validate-infrastructure-cluster-x-k8s-io-v1alpha5-openstackmachine" "UID"="d5aa80fb-a04d-4615-a58b-ef4edde0c0d8" "kind"={"group":"infrastructure.cluster.x-k8s.io","version":"v1alpha5","kind":"OpenStackMachine"} "resource"={"group":"infrastructure.cluster.x-k8s.io","version":"v1alpha5","resource":"openstackmachines"}

I0615 07:39:49.589676       1 http.go:136] controller-runtime/webhook/webhooks "msg"="wrote response" "code"=200 "reason"="" "webhook"="/validate-infrastructure-cluster-x-k8s-io-v1alpha5-openstackmachine" "UID"="d5aa80fb-a04d-4615-a58b-ef4edde0c0d8" "allowed"=true

I0615 07:39:49.614983       1 openstackmachine_controller.go:287] controller/openstackmachine "msg"="Cluster infrastructure is not ready yet, requeuing machine" "cluster"="capi-quickstart" "machine"="capi-quickstart-md-0-7674b8c4d7-2ngwq" "name"="capi-quickstart-md-0-g9wdj" "namespace"="default" "openStackCluster"="capi-quickstart" "openStackMachine"="capi-quickstart-md-0-g9wdj" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="OpenStackMachine"

I0615 07:39:49.664738       1 http.go:97] controller-runtime/webhook/webhooks "msg"="received request" "webhook"="/mutate-infrastructure-cluster-x-k8s-io-v1alpha5-openstackmachine" "UID"="2755c0f0-f93b-4c81-ae3a-3b51aa0c8864" "kind"={"group":"infrastructure.cluster.x-k8s.io","version":"v1alpha5","kind":"OpenStackMachine"} "resource"={"group":"infrastructure.cluster.x-k8s.io","version":"v1alpha5","resource":"openstackmachines"}

I0615 07:39:49.665887       1 http.go:136] controller-runtime/webhook/webhooks "msg"="wrote response" "code"=200 "reason"="" "webhook"="/mutate-infrastructure-cluster-x-k8s-io-v1alpha5-openstackmachine" "UID"="2755c0f0-f93b-4c81-ae3a-3b51aa0c8864" "allowed"=true

I0615 07:39:49.683998       1 openstackmachine_controller.go:287] controller/openstackmachine "msg"="Cluster infrastructure is not ready yet, requeuing machine" "cluster"="capi-quickstart" "machine"="capi-quickstart-md-0-7674b8c4d7-vstqf" "name"="capi-quickstart-md-0-5fd4j" "namespace"="default" "openStackCluster"="capi-quickstart" "openStackMachine"="capi-quickstart-md-0-5fd4j" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="OpenStackMachine"

I0615 07:39:49.744604       1 openstackmachine_controller.go:287] controller/openstackmachine "msg"="Cluster infrastructure is not ready yet, requeuing machine" "cluster"="capi-quickstart" "machine"="capi-quickstart-md-0-7674b8c4d7-m7sbl" "name"="capi-quickstart-md-0-nxmsz" "namespace"="default" "openStackCluster"="capi-quickstart" "openStackMachine"="capi-quickstart-md-0-nxmsz" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="OpenStackMachine"

I0615 07:39:49.774955       1 openstackmachine_controller.go:287] controller/openstackmachine "msg"="Cluster infrastructure is not ready yet, requeuing machine" "cluster"="capi-quickstart" "machine"="capi-quickstart-md-0-7674b8c4d7-2ngwq" "name"="capi-quickstart-md-0-g9wdj" "namespace"="default" "openStackCluster"="capi-quickstart" "openStackMachine"="capi-quickstart-md-0-g9wdj" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="OpenStackMachine"

I0615 07:39:49.787419       1 openstackmachine_controller.go:202] controller/openstackmachine "msg"="Reconciling Machine delete" "cluster"="capi-quickstart" "machine"="capi-quickstart-md-0-7674b8c4d7-vstqf" "name"="capi-quickstart-md-0-5fd4j" "namespace"="default" "openStackCluster"="capi-quickstart" "openStackMachine"="capi-quickstart-md-0-5fd4j" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="OpenStackMachine"

I0615 07:39:49.892724       1 openstackmachine_controller.go:202] controller/openstackmachine "msg"="Reconciling Machine delete" "cluster"="capi-quickstart" "machine"="capi-quickstart-md-0-7674b8c4d7-2ngwq" "name"="capi-quickstart-md-0-g9wdj" "namespace"="default" "openStackCluster"="capi-quickstart" "openStackMachine"="capi-quickstart-md-0-g9wdj" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="OpenStackMachine"

I0615 07:39:49.910477       1 openstackmachine_controller.go:202] controller/openstackmachine "msg"="Reconciling Machine delete" "cluster"="capi-quickstart" "machine"="capi-quickstart-md-0-7674b8c4d7-m7sbl" "name"="capi-quickstart-md-0-nxmsz" "namespace"="default" "openStackCluster"="capi-quickstart" "openStackMachine"="capi-quickstart-md-0-nxmsz" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="OpenStackMachine"

panic: runtime error: invalid memory address or nil pointer dereference

[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x14e98c6]

 

goroutine 406 [running]:

sigs.k8s.io/cluster-api-provider-openstack/controllers.machineToInstanceSpec(0xc00026f400, 0xc0003f4000, 0xc000306280, {0x0, 0x0})

        /workspace/controllers/openstackmachine_controller.go:474 +0x546

sigs.k8s.io/cluster-api-provider-openstack/controllers.(*OpenStackMachineReconciler).reconcileDelete(0x1a6c338, {0x1a6c338, 0xc000be3020}, 0xc000781680, 0xc000306280, 0xc0008d81c0, 0xc00026f400, 0xc0003f4000, 0xc000306280)

        /workspace/controllers/openstackmachine_controller.go:252 +0x46b

sigs.k8s.io/cluster-api-provider-openstack/controllers.(*OpenStackMachineReconciler).Reconcile(0xc000678240, {0x1a6c338, 0xc000be3020}, {{{0xc000d10060, 0x1766dc0}, {0xc0011a0c20, 0x30}}})

        /workspace/controllers/openstackmachine_controller.go:157 +0xb05

sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0xc0004bc0b0, {0x1a6c338, 0xc000be2ff0}, {{{0xc000d10060, 0x1766dc0}, {0xc0011a0c20, 0x413c34}}})

        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:114 +0x26f

sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc0004bc0b0, {0x1a6c290, 0xc000565b80}, {0x16a4c00, 0xc0007dcec0})

        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:311 +0x33e

sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc0004bc0b0, {0x1a6c290, 0xc000565b80})

        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:266 +0x205

sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()

        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:227 +0x85

created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2

        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:223 +0x357


```

**What did you expect to happen:**
the CAPO cluster to be completely delete

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]


**Environment:**

- Cluster API Provider OpenStack version (Or `git rev-parse HEAD` if manually built):
- Cluster-API version: V1/alpha5
- OpenStack version: stein
- Minikube/KIND version:
- Kubernetes version (use `kubectl version`): 1.22.9
- OS (e.g. from `/etc/os-release`): ubuntu 20.04

Same issue here:
```
I0622 13:46:17.042527       1 openstackmachine_controller.go:287] controller/openstackmachine "msg"="Cluster infrastructure is not ready yet, requeuing machine" "cluster"="capi-quickstart" "machine"="capi-quickstart-md-0-547f548bb7-wkhgd" "name"="capi-quickstart-md-0-c6j9b" "namespace"="default" "openStackCluster"="capi-quickstart" "openStackMachine"="capi-quickstart-md-0-c6j9b" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="OpenStackMachine" 
I0622 13:46:17.043654       1 openstackmachine_controller.go:287] controller/openstackmachine "msg"="Cluster infrastructure is not ready yet, requeuing machine" "cluster"="capi-quickstart" "machine"="capi-quickstart-md-0-547f548bb7-d9shh" "name"="capi-quickstart-md-0-ngb46" "namespace"="default" "openStackCluster"="capi-quickstart" "openStackMachine"="capi-quickstart-md-0-ngb46" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="OpenStackMachine" 
I0622 13:46:17.053266       1 openstackmachine_controller.go:287] controller/openstackmachine "msg"="Cluster infrastructure is not ready yet, requeuing machine" "cluster"="capi-quickstart" "machine"="capi-quickstart-md-0-547f548bb7-gxtg8" "name"="capi-quickstart-md-0-7bg4q" "namespace"="default" "openStackCluster"="capi-quickstart" "openStackMachine"="capi-quickstart-md-0-7bg4q" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="OpenStackMachine" 
I0622 13:46:17.436218       1 openstackmachine_controller.go:202] controller/openstackmachine "msg"="Reconciling Machine delete" "cluster"="capi-quickstart" "machine"="capi-quickstart-md-0-547f548bb7-wkhgd" "name"="capi-quickstart-md-0-c6j9b" "namespace"="default" "openStackCluster"="capi-quickstart" "openStackMachine"="capi-quickstart-md-0-c6j9b" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="OpenStackMachine" 
I0622 13:46:17.442631       1 openstackmachine_controller.go:202] controller/openstackmachine "msg"="Reconciling Machine delete" "cluster"="capi-quickstart" "machine"="capi-quickstart-md-0-547f548bb7-d9shh" "name"="capi-quickstart-md-0-ngb46" "namespace"="default" "openStackCluster"="capi-quickstart" "openStackMachine"="capi-quickstart-md-0-ngb46" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="OpenStackMachine" 
I0622 13:46:17.451274       1 openstackmachine_controller.go:202] controller/openstackmachine "msg"="Reconciling Machine delete" "cluster"="capi-quickstart" "machine"="capi-quickstart-md-0-547f548bb7-gxtg8" "name"="capi-quickstart-md-0-7bg4q" "namespace"="default" "openStackCluster"="capi-quickstart" "openStackMachine"="capi-quickstart-md-0-7bg4q" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="OpenStackMachine" 
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x14e98c6]

goroutine 432 [running]:
sigs.k8s.io/cluster-api-provider-openstack/controllers.machineToInstanceSpec(0xc000581c00, 0xc000975b00, 0xc000cbe280, {0x0, 0x0})
	/workspace/controllers/openstackmachine_controller.go:474 +0x546
sigs.k8s.io/cluster-api-provider-openstack/controllers.(*OpenStackMachineReconciler).reconcileDelete(0x1a6c338, {0x1a6c338, 0xc000400ba0}, 0xc0002bd5c0, 0xc000cbe280, 0xc0008408c0, 0xc000581c00, 0xc000975b00, 0xc000cbe280)
	/workspace/controllers/openstackmachine_controller.go:252 +0x46b
sigs.k8s.io/cluster-api-provider-openstack/controllers.(*OpenStackMachineReconciler).Reconcile(0xc0005562d0, {0x1a6c338, 0xc000400ba0}, {{{0xc0008269e6, 0x1766dc0}, {0xc0006ee140, 0x30}}})
	/workspace/controllers/openstackmachine_controller.go:157 +0xb05
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0xc00014d600, {0x1a6c338, 0xc000400b70}, {{{0xc0008269e6, 0x1766dc0}, {0xc0006ee140, 0x413c34}}})
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:114 +0x26f
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc00014d600, {0x1a6c290, 0xc00003b280}, {0x16a4c00, 0xc000362a20})
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:311 +0x33e
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc00014d600, {0x1a6c290, 0xc00003b280})
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:266 +0x205
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:227 +0x85
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:223 +0x357
```

We're hitting the same problem;

```log
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x14e9b06]

goroutine 454 [running]:
sigs.k8s.io/cluster-api-provider-openstack/controllers.machineToInstanceSpec(0xc000a33c00, 0xc000c71680, 0xc00039a280, {0x0, 0x0})
	/workspace/controllers/openstackmachine_controller.go:474 +0x546
sigs.k8s.io/cluster-api-provider-openstack/controllers.(*OpenStackMachineReconciler).reconcileDelete(0x1a6c438, {0x1a6c438, 0xc000c6b6b0}, 0xc000834f40, 0xc00039a280, 0xc000c94000, 0xc000a33c00, 0xc000c71680, 0xc00039a280)
	/workspace/controllers/openstackmachine_controller.go:252 +0x46b
sigs.k8s.io/cluster-api-provider-openstack/controllers.(*OpenStackMachineReconciler).Reconcile(0xc000454150, {0x1a6c438, 0xc000c6b6b0}, {{{0xc0007e3410, 0x1766dc0}, {0xc0007b7dc0, 0x30}}})
	/workspace/controllers/openstackmachine_controller.go:157 +0xb05
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0xc0000d4370, {0x1a6c438, 0xc000c6b680}, {{{0xc0007e3410, 0x1766dc0}, {0xc0007b7dc0, 0x413c34}}})
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:114 +0x26f
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc0000d4370, {0x1a6c390, 0xc000540c00}, {0x16a4c00, 0xc000adec20})
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:311 +0x33e
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc0000d4370, {0x1a6c390, 0xc000540c00})
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:266 +0x205
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:227 +0x85
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:223 +0x357
```

This has started after I created a cluster and immediately delete it, I suspect that's got something to do with it

--- a/controllers/openstackcluster_controller.go
+++ b/controllers/openstackcluster_controller.go
@@ -231,8 +231,8 @@ func deleteBastion(scope *scope.Scope, cluster *clusterv1.Cluster, openStackClus
 			}
 		}
 
-		instanceSpec := bastionToInstanceSpec(openStackCluster, cluster.Name)
-		if err = computeService.DeleteInstance(openStackCluster, instanceSpec, instanceStatus); err != nil {
+		rootVolume := openStackCluster.Spec.Bastion.Instance.RootVolume
+		if err = computeService.DeleteInstance(openStackCluster, instanceStatus, instanceName, rootVolume); err != nil {
 			handleUpdateOSCError(openStackCluster, errors.Errorf("failed to delete bastion: %v", err))
 			return errors.Errorf("failed to delete bastion: %v", err)
 		}

--- a/controllers/openstackmachine_controller.go
+++ b/controllers/openstackmachine_controller.go
@@ -278,15 +278,7 @@ func (r *OpenStackMachineReconciler) reconcileDelete(ctx context.Context, scope 
 		}
 	}
 
-	instanceSpec, err := machineToInstanceSpec(openStackCluster, machine, openStackMachine, "")
-	if err != nil {
-		err = errors.Errorf("machine spec is invalid: %v", err)
-		handleUpdateMachineError(scope.Logger, openStackMachine, err)
-		conditions.MarkFalse(openStackMachine, infrav1.InstanceReadyCondition, infrav1.InvalidMachineSpecReason, clusterv1.ConditionSeverityError, err.Error())
-		return ctrl.Result{}, err
-	}
-
-	if err := computeService.DeleteInstance(openStackMachine, instanceSpec, instanceStatus); err != nil {
+	if err := computeService.DeleteInstance(openStackMachine, instanceStatus, openStackMachine.Name, openStackMachine.Spec.RootVolume); err != nil {
 		handleUpdateMachineError(scope.Logger, openStackMachine, errors.Errorf("error deleting OpenStack instance %s with ID %s: %v", instanceStatus.Name(), instanceStatus.ID(), err))
 		conditions.MarkFalse(openStackMachine, infrav1.InstanceReadyCondition, infrav1.InstanceDeleteFailedReason, clusterv1.ConditionSeverityError, "Deleting instance failed: %v", err)
 		return ctrl.Result{}, nil

--- a/pkg/cloud/services/compute/instance.go
+++ b/pkg/cloud/services/compute/instance.go
@@ -513,7 +513,7 @@ func (s *Service) GetManagementPort(openStackCluster *infrav1.OpenStackCluster, 
 	return &allPorts[0], nil
 }
 
-func (s *Service) DeleteInstance(eventObject runtime.Object, instanceSpec *InstanceSpec, instanceStatus *InstanceStatus) error {
+func (s *Service) DeleteInstance(eventObject runtime.Object, instanceStatus *InstanceStatus, instanceName string, rootVolume *infrav1.RootVolume) error {
 	if instanceStatus == nil {
 		/*
 			We create a boot-from-volume instance in 2 steps:
@@ -533,9 +533,8 @@ func (s *Service) DeleteInstance(eventObject runtime.Object, instanceSpec *Insta
 			Note that we don't need to separately delete the root volume when deleting the instance because
 			DeleteOnTermination will ensure it is deleted in that case.
 		*/
-		rootVolume := instanceSpec.RootVolume
 		if hasRootVolume(rootVolume) {
-			name := rootVolumeName(instanceSpec.Name)
+			name := rootVolumeName(instanceName)
 			volume, err := s.getVolumeByName(name)
 			if err != nil {
 				return err

--- a/pkg/cloud/services/compute/instance_test.go
+++ b/pkg/cloud/services/compute/instance_test.go
@@ -1031,15 +1031,14 @@ func TestService_DeleteInstance(t *testing.T) {
 	tests := []struct {
 		name           string
 		eventObject    runtime.Object
-		instanceSpec   func() *InstanceSpec
 		instanceStatus func() *InstanceStatus
+		rootVolume     *infrav1.RootVolume
 		expect         func(computeRecorder *MockClientMockRecorder, networkRecorder *mock_networking.MockNetworkClientMockRecorder)
 		wantErr        bool
 	}{
 		{
 			name:           "Defaults",
 			eventObject:    &infrav1.OpenStackMachine{},
-			instanceSpec:   getDefaultInstanceSpec,
 			instanceStatus: getDefaultInstanceStatus,
 			expect: func(computeRecorder *MockClientMockRecorder, networkRecorder *mock_networking.MockNetworkClientMockRecorder) {
 				computeRecorder.ListAttachedInterfaces(instanceUUID).Return([]attachinterfaces.Interface{
@@ -1063,16 +1062,12 @@ func TestService_DeleteInstance(t *testing.T) {
 			wantErr: false,
 		},
 		{
-			name:        "Dangling volume",
-			eventObject: &infrav1.OpenStackMachine{},
-			instanceSpec: func() *InstanceSpec {
-				spec := getDefaultInstanceSpec()
-				spec.RootVolume = &infrav1.RootVolume{
-					Size: 50,
-				}
-				return spec
-			},
+			name:           "Dangling volume",
+			eventObject:    &infrav1.OpenStackMachine{},
 			instanceStatus: func() *InstanceStatus { return nil },
+			rootVolume: &infrav1.RootVolume{
+				Size: 50,
+			},
 			expect: func(computeRecorder *MockClientMockRecorder, networkRecorder *mock_networking.MockNetworkClientMockRecorder) {
 				// Fetch volume by name
 				volumeName := fmt.Sprintf("%s-root", openStackMachineName)
@@ -1112,7 +1107,7 @@ func TestService_DeleteInstance(t *testing.T) {
 					"", mockNetworkClient, logr.Discard(),
 				),
 			}
-			if err := s.DeleteInstance(tt.eventObject, tt.instanceSpec(), tt.instanceStatus()); (err != nil) != tt.wantErr {
+			if err := s.DeleteInstance(tt.eventObject, tt.instanceStatus(), openStackMachineName, tt.rootVolume); (err != nil) != tt.wantErr {
 				t.Errorf("Service.DeleteInstance() error = %v, wantErr %v", err, tt.wantErr)
 			}
 		})

